### PR TITLE
Switch StripeConnectAccountController away from ja_resource and canary

### LINKED ISSUE #905: Switch StripeConnectAccountController away from ja_resource and canary
Progress on #864 

# Problem

We want to move away from the `ja_resource` and `canary` combo used in most of our controller modules. The central issue for this project is #864

There is a generic process of how to do it in that issue, as well as tracking for the milestone progress. 

The procedure is also pasted below. If this specific controller deviates from this standard procedure, feel free to ask for help here, or by opening a PR.

The PR should indicate that it's closing this issue and is progressing #864 further along.

# Procedure

- pick a controller to switch
- view actions this controller should support in `lib/code_corps_web/router.ex`
- add `action_fallback CodeCorpsWeb.FallbackController`. This makes it so in case a with clause within the controller action fails, a sort of "catch all" function will be called from `FallbackController`
- add `plug CodeCorpsWeb.Plug.DataToAttributes`. This converts the `data` hash within the conn params from the json api structure into a flat map convenient for usage with changesets
- add `plug CodeCorpsWeb.Plug.IdsToIntegers`. This casts all id parameters into integers, since this is what our authorization layer and Phoenix in general expects, while ember is sending strings.
- for each action
   - remove that action from any `load_resource`, `load_and_authorize_resource`, etc. plugs in that controller
   - remove any `def handle_#{action}` functions
   - write a `def #{action}(conn, params) do`. For the standard `create`, `update` `delete`, `show` and index actions. See examples below on how these functions should be written, usually
   - rewrite the policy function for that action to support params instead of a changeset if necessary. See section below for pointers.

That should be the whole procedure. We already have controller tests in place, so if the process has been done correctly, the tests should pass.

Some additional steps might be necessary, however, if there are some specifics to an action. For those, we will be here to provide feedback.

## How to rewrite controller actions

```Elixir
@spec index(Conn.t, map) :: Conn.t
def index(%Conn{} = conn, %{} = params) do
  with resources <- ResourceModule |> Query.id_filter(params) |> Repo.all do
    conn |> render("index.json-api", data: resources)
  end
end

@spec show(Conn.t, map) :: Conn.t
def show(%Conn{} = conn, %{"id" => id}) do
  with %ResourceModule{} = resource <- ResourceModule |> Repo.get(id) do
    conn |> render("show.json-api", data: resource)
  end
end

@spec create(Plug.Conn.t, map) :: Conn.t
def create(%Conn{} = conn, %{} = params) do
  with %User{} = current_user <- conn |> Guardian.Plug.current_resource,
       {:ok, :authorized} <- current_user |> Policy.authorize(:create, params),
       {:ok, %Comment{} = resource} <- %ResourceModule{} |> ResourceModule.create_changeset(params) |> Repo.insert do
    conn |> put_status(:created) |> render("show.json-api", data: resource)
  end
end

@spec update(Conn.t, map) :: Conn.t
def update(%Conn{} = conn, %{"id" => id} = params) do
  with %ResourceModule{} = resource <- ResourceModule |> Repo.get(id),
       %User{} = current_user <- conn |> Guardian.Plug.current_resource,
       {:ok, :authorized} <- current_user |> Policy.authorize(:update, resource),
       {:ok, %ResourceModule{} = resource} <- resource |> ResourceModule.changeset(params) |> Repo.update do
    conn |> render("show.json-api", data: resource)
  end
end
```

So to recap, the actions work like

### `index` action

- apply filters from query params to load data
- render data

### `show` action

- load resource
- render resource

### `create` action

- authorize action for current user and parameters (requires rewriting that policy to use a params map instead of a changeset, expect `{:ok, :authorized}`
- perform insert action
- render resulting resource

### `update` and `delete` actions

- load resource
- authorize action for current user on resource, expect `{:ok, :authorized}`
- perform update/delete
- render resource, or no content for delete

### Examples in code

-`CodeCorpsWeb.CommentController` has examples for standard `index`, `show`, `create` and `update` actions

## How to rewrite policies

This consists of several relatively simple steps.

- For each action being switch away, we need to move the `def can` function clause out of the `defimpl` block in `policy.ex` into the root namespace of the `CodeCorps.Policy` module. 
- We also need to make the clause private instead of public.

- If any of the clauses expect a changeset, they should now expect a simple map of params:

```Elixir
defp can?(%User{} = user, :some_action, %Changeset{data: %Resource{}} = changeset), do: Policy.Resource.some_action?(user, changeset)
```

becomes

```Elixir
defp can?(%User{} = user, :some_action, %{} = params), do: Policy.Resource.some_action?(user, params)
```

- in this case, the policy module itself also need to be rewritten to support a params map. Since in most cases, these policies simply look at the `changeset.changes` map, the rewrite ought to be simpler. If there is trouble, someone will be available to provide pointers.

### Examples in code

- the rewrite has been done for `Policy.Comment.create?`

### PR DESCRIPTION
**Pull request by treble37**
Closes #905

Removes reliance and ja_resources and canary 

## References
Fixes # 905

Progress on: # 865

### CODE REVIEW

**Comment by treble37** [comment]
I think I need some help for the 2 tests that are failing locally...

File: lib/code_corps_web/controllers/stripe_connect_account_controller.ex [hunk 1]
```diff
@@ -1,27 +1,47 @@
...
+  plug CodeCorpsWeb.Plug.DataToAttributes
 
   @spec model :: module
   def model, do: CodeCorps.StripeConnectAccount
 
-  def handle_create(conn, attributes) do
-    attributes
-    |> Map.put("tos_acceptance_ip", conn |> ConnUtils.extract_ip)
-    |> Map.put("tos_acceptance_user_agent", conn |> ConnUtils.extract_user_agent)
-    |> Map.put("managed", true)
```
Comment (begedin) [inline lib/code_corps_web/controllers/stripe_connect_account_controller.ex]: You need to do the same to `params` below, to get this passing correctly.

File: lib/code_corps/policy/policy.ex [hunk 2]
```diff
@@ -51,6 +51,9 @@ defmodule CodeCorps.Policy do
   defp can?(%User{} = current_user, :delete, %UserSkill{} = user_skill, %{}), do: Policy.UserSkill.delete?(current_user, user_skill)
   defp can?(%User{} = current_user, :create, %UserRole{} = user_role, %{}), do: Policy.UserRole.create?(current_user, user_role)
   defp can?(%User{} = current_user, :delete, %UserRole{} = user_role, %{}), do: Policy.UserRole.delete?(current_user, user_role)
+  defp can?(%User{} = user, :show, %StripeConnectAccount{} = stripe_connect_account, params), do: Policy.StripeConnectAccount.show?(user, params)
```
Comment (begedin) [inline lib/code_corps/policy/policy.ex]: This should be 

```Elixir
defp can?(%User{} = user, :show, %StripeConnectAccount{} = stripe_connect_account, %{}), do: Policy.StripeConnectAccount.show?(user, stripe_connect_account)
```

Once you make that change, you can remove the equivalent from line 97

File: lib/code_corps/policy/policy.ex [hunk 3]
```diff
@@ -51,6 +51,9 @@ defmodule CodeCorps.Policy do
...
+  defp can?(%User{} = user, :create, %StripeConnectAccount{}, params), do: Policy.StripeConnectAccount.create?(user, params)
```
Comment (begedin) [inline lib/code_corps/policy/policy.ex]: `%{} = params`, so we typecheck against a map.

File: lib/code_corps/policy/policy.ex [hunk 4]
```diff
@@ -51,6 +51,9 @@ defmodule CodeCorps.Policy do
...
+  defp can?(%User{} = user, :update, %StripeConnectAccount{} = stripe_connect_account, params), do: Policy.StripeConnectAccount.update?(user, stripe_connect_account)
```
Comment (begedin) [inline lib/code_corps/policy/policy.ex]: `%{}` here, since we aren't using the params, but still want to typecheck.

You should be getting compiler warnings for these, so be sure to get those eliminated.

File: lib/code_corps/policy/policy.ex [hunk 5]
```diff
@@ -92,7 +95,7 @@ defmodule CodeCorps.Policy do
     def can?(%User{} = user, :create, Skill), do: Policy.Skill.create?(user)
 
     def can?(%User{} = user, :show, %StripeConnectAccount{} = stripe_connect_account), do: Policy.StripeConnectAccount.show?(user, stripe_connect_account)
-    def can?(%User{} = user, :create, %Changeset{ data: %StripeConnectAccount{}} = changeset), do: Policy.StripeConnectAccount.create?(user, changeset)
+    def can?(%User{} = user, :create, %StripeConnectAccount{}, params), do: Policy.StripeConnectAccount.create?(user, params)
     def can?(%User{} = user, :update, %StripeConnectAccount{} = stripe_connect_account), do: Policy.StripeConnectAccount.update?(user, stripe_connect_account)
```
Comment (begedin) [inline lib/code_corps/policy/policy.ex]: The end result is that you want to remove lines 97, 98 and 99 in favor of lines 54, 55 and 56

File: lib/code_corps/policy/stripe_connect_account.ex [hunk 6]
```diff
@@ -3,11 +3,11 @@ defmodule CodeCorps.Policy.StripeConnectAccount do
 
   alias CodeCorps.{StripeConnectAccount, User}
 
-  def show?(%User{} = user, %StripeConnectAccount{} = stripe_connect_account),
-    do: stripe_connect_account |> get_organization() |> owned_by?(user)
+  def show?(%User{} = user, params),
+    do: params |> get_organization() |> owned_by?(user)
```
Comment (begedin) [inline lib/code_corps/policy/stripe_connect_account.ex]: This was an unnecessary change here. `show?` should still accept an account as a second argument.

File: lib/code_corps/policy/stripe_connect_account.ex [hunk 7]
```diff
@@ -3,11 +3,11 @@ defmodule CodeCorps.Policy.StripeConnectAccount do
...
 
-  def create?(%User{} = user, %Ecto.Changeset{} = changeset),
-    do: changeset |> get_organization() |> owned_by?(user)
+  def create?(%User{} = user, params),
```
Comment (begedin) [inline lib/code_corps/policy/stripe_connect_account.ex]: `%{} = params` for the same reason as above.

File: lib/code_corps_web/controllers/stripe_connect_account_controller.ex [hunk 8]
```diff
@@ -1,27 +1,47 @@
...
   alias CodeCorps.ConnUtils
   alias CodeCorps.StripeConnectAccount
   alias CodeCorps.StripeService.StripeConnectAccountService
+  alias CodeCorps.User
 
-  plug :load_and_authorize_changeset, model: StripeConnectAccount, only: [:create]
-  plug :load_and_authorize_resource, model: StripeConnectAccount, only: [:show, :update]
-  plug JaResource
+  action_fallback CodeCorpsWeb.FallbackController
+  plug CodeCorpsWeb.Plug.DataToAttributes
```
Comment (begedin) [inline lib/code_corps_web/controllers/stripe_connect_account_controller.ex:10]: don't forget `plug CodeCorpsWeb.IdsToIntegers` to

**Review by begedin — CHANGES_REQUESTED** [review]

File: lib/code_corps_web/controllers/stripe_connect_account_controller.ex (hunk 8 above)
Comment (treble37) [reply lib/code_corps_web/controllers/stripe_connect_account_controller.ex:10]: I think I fixed all the issues now, thanks @begedin

File: lib/code_corps/policy/policy.ex [hunk 9]
```diff
@@ -51,6 +51,9 @@ defmodule CodeCorps.Policy do
   defp can?(%User{} = current_user, :delete, %UserSkill{} = user_skill, %{}), do: Policy.UserSkill.delete?(current_user, user_skill)
   defp can?(%User{} = current_user, :create, %UserRole{} = user_role, %{}), do: Policy.UserRole.create?(current_user, user_role)
   defp can?(%User{} = current_user, :delete, %UserRole{} = user_role, %{}), do: Policy.UserRole.delete?(current_user, user_role)
+  defp can?(%User{} = user, :show, %StripeConnectAccount{} = stripe_connect_account, %{} = params), do: Policy.StripeConnectAccount.show?(user, stripe_connect_account)
```
Comment (begedin) [inline lib/code_corps/policy/policy.ex]: `%{} = params` should throw out a compile warning when you run the tests, because `params` is not being used. Is it not happening for you. `%{}` should be correct.

File: lib/code_corps/policy/policy.ex [hunk 10]
```diff
@@ -51,6 +51,9 @@ defmodule CodeCorps.Policy do
...
+  defp can?(%User{} = user, :create, %StripeConnectAccount{}, %{} = params), do: Policy.StripeConnectAccount.create?(user, params)
+  defp can?(%User{} = user, :update, %StripeConnectAccount{} = stripe_connect_account, %{} = params), do: Policy.StripeConnectAccount.update?(user, stripe_connect_account)
```
Comment (begedin) [inline lib/code_corps/policy/policy.ex]: Same here with `%{} = params` -> `%{}`

**Review by begedin — DISMISSED** [review]
Just two more very minor details and this is good to go. Good work @treble37

**Comment by begedin** [comment]
Another reminder to rebase and squash as you add the fixes @treble37, not that you haven't been doing that correctly up to this point. I just find it's easier to remind, than to wait another day (due to time difference) to merge something that's otherwise 👍

File: lib/code_corps/policy/policy.ex (hunk 9 above)
Comment (treble37) [reply lib/code_corps/policy/policy.ex]: @begedin - I think I missed it because it throws out warnings only when I open a terminal and do `mix test` the first time....subsequent runs of mix test don't show the `params is unused warning`.....

**Comment by treble37** [comment]
Thanks for the kind reminder, should be rebased and squashed, thanks!

**Review by joshsmith — APPROVED** [review]
